cpp/src/arrow/filesystem/hdfs.cc

```diff
@@ -0,0 +1,413 @@
+          select.base_dir, "' is a file");
+    }
+    RETURN_NOT_OK(StatSelector(wd, select.base_dir, select, 0, out));
+    return Status::OK();
+  }
+
+  Status CreateDir(const std::string& path, bool recursive) {
+    if (IsDirectory(path)) {
+      return Status::OK();
+    }
+    if (!recursive) {
+      const auto parent = GetAbstractPathParent(path).first;
+      if (!parent.empty() && !IsDirectory(parent)) {
+        return Status::IOError("Cannot create directory '", path,
+                               "': parent is not a directory");
+      }
+    }
+    RETURN_NOT_OK(client_->MakeDirectory(path));
+    return Status::OK();
+  }
+
+  Status DeleteDir(const std::string& path) {
+    if (!IsDirectory(path)) {
+      return Status::IOError("Cannot delete directory '", path, "': not a directory");
+    }
+    RETURN_NOT_OK(client_->DeleteDirectory(path));
+    return Status::OK();
+  }
+
+  Status DeleteDirContents(const std::string& path) {
+    std::vector<std::string> file_list;
+    RETURN_NOT_OK(client_->GetChildren(path, &file_list));
+    for (auto file : file_list) {
+      RETURN_NOT_OK(client_->Delete(file, /*recursive=*/true));
+    }
+    return Status::OK();
+  }
+
+  Status DeleteFile(const std::string& path) {
+    if (IsDirectory(path)) {
+      return Status::IOError("path is a directory");
+    }
+    RETURN_NOT_OK(client_->Delete(path));
+    return Status::OK();
+  }
+
+  Status Move(const std::string& src, const std::string& dest) {
+    RETURN_NOT_OK(client_->Rename(src, dest));
+    return Status::OK();
+  }
+
+  Status CopyFile(const std::string& src, const std::string& dest) {
+    // TODO implement this (but only if HDFS supports on-server copy)
+    return Status::NotImplemented("HadoopFileSystem::CopyFile is not supported yet");
+  }
+
+  Status OpenInputStream(const std::string& path, std::shared_ptr<io::InputStream>* out) {
+    std::shared_ptr<io::HdfsReadableFile> file;
+    RETURN_NOT_OK(client_->OpenReadable(path, &file));
+    *out = file;
+    return Status::OK();
+  }
+
+  Status OpenInputFile(const std::string& path,
+                       std::shared_ptr<io::RandomAccessFile>* out) {
+    std::shared_ptr<io::HdfsReadableFile> file;
+    RETURN_NOT_OK(client_->OpenReadable(path, &file));
+    *out = file;
+    return Status::OK();
+  }
+
+  Status OpenOutputStream(const std::string& path,
+                          std::shared_ptr<io::OutputStream>* out) {
+    bool append = false;
+    return OpenOutputStreamGeneric(path, append, out);
+  }
+
+  Status OpenAppendStream(const std::string& path,
+                          std::shared_ptr<io::OutputStream>* out) {
+    bool append = true;
+    return OpenOutputStreamGeneric(path, append, out);
+  }
+
+ protected:
+  HdfsOptions options_;
+  std::shared_ptr<::arrow::io::HadoopFileSystem> client_;
+
+  void PathInfoToFileStats(const io::HdfsPathInfo& info, FileStats* out) {
```

Comment:
This function should take the path as parameter and return FileStats directly, would make thing more natural.

Reply:
No, because it acts on an existing `FileStats`. I mean other members have been set by the caller.